lib/ansible/modules/crypto/openssl_pkcs12.py

```diff
@@ -237,7 +237,54 @@ def _check_pkey_passphrase():
         if not state_and_perms:
             return state_and_perms
 
-        return _check_pkey_passphrase
+        if os.path.exists(self.path) and module.params['action'] == 'export':
+            dummy = self.generate(module)
+            self.src = self.path
+            try:
+                pkcs12_privatekey, pkcs12_certificate, pkcs12_other_certificates, pkcs12_friendly_name = self.parse()
+            except crypto.Error:
+                return False
+            if (pkcs12_privatekey is not None) and (self.privatekey_path is not None):
+                expected_pkey = crypto.dump_privatekey(crypto.FILETYPE_PEM,
+                                                       self.pkcs12.get_privatekey())
+                if pkcs12_privatekey != expected_pkey:
+                    return False
+            elif bool(pkcs12_privatekey) != bool(self.privatekey_path):
+                return False
+
+            if (pkcs12_certificate is not None) and (self.certificate_path is not None):
+
+                expected_cert = crypto.dump_certificate(crypto.FILETYPE_PEM,
+                                                        self.pkcs12.get_certificate())
+                if pkcs12_certificate != expected_cert:
+                    for pkcs_cert in pkcs12_other_certificates:
+                        if pkcs_cert == expected_cert:
+                            break
+                    else:
+                        return False
+            elif bool(pkcs12_certificate) != bool(self.certificate_path):
+                return False
+
+            if (pkcs12_other_certificates is not None) and (self.ca_certificates is not None):
+                expected_ca_certs = [crypto.dump_certificate(crypto.FILETYPE_PEM,
+                                                             ca_cert) for ca_cert in self.pkcs12.get_ca_certificates()]
+                expected_ca_certs.append(crypto.dump_certificate(crypto.FILETYPE_PEM, self.pkcs12.get_certificate()))
+                pkcs12_other_certificates.append(pkcs12_certificate)
+
+                if set(pkcs12_other_certificates) != set(expected_ca_certs):
+                    return False
+            elif bool(pkcs12_other_certificates) != bool(self.ca_certificates):
+                return False
+
+            if not pkcs12_other_certificates:
```

Comment:
What is this `if` for?

Reply:
It appears that if the pkcs12 file has additional certs, the `get_friendlyname` method ( from `parse` ) returns None, making this fail unnecessarily. That said, I guess I could/should change this to be 
```python
not pkcs12_other_certificates and not pkcs12_friendly_name
```

Reply:
Hmm, that's even more strange. I'd say outright terrible :)

I've looked at the pyOpenSSL implementation of encoding and decoding PKCS12 files, it is essentially calling the OpenSSL functions from the other comment. The friendly name comes from the main cert extracted from the PKCS12 file, i.e. when the order is f**ked up, I would expect `friendly_name` to also be wrong.